bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java

```diff
@@ -140,22 +163,34 @@ public void stopWithoutFailureLogging()
     public void stop()
             throws LifeCycleStopException
     {
-        if (!stopInternal(true).isEmpty()) {
-            // No suppressed exception propagation, exceptions already logged
+        final class LogEachAndDiscardFailureHandler
+                implements LifeCycleStopFailureHandler
+        {
+            private boolean exceptionOccurred;
+
+            @Override
+            public void handlePreDestroyException(Class<?> klass, Method method, Exception exception)
+            {
+                log.error(exception, "Exception in PreDestroy method %s::%s()", klass.getName(), method.getName());
+                this.exceptionOccurred = true;
+            }
+        }
+
+        if (stop(new LogEachAndDiscardFailureHandler()).exceptionOccurred) {
             throw new LifeCycleStopException();
         }
     }
 
     /**
      * Stop the life cycle - all instances will have their {@link javax.annotation.PreDestroy} method(s) called
-     * and any exceptions raised will be returned.
+     * and any exceptions raised will be passed to the provided {@link LifeCycleStopFailureHandler} to collect.
      *
-     * @return All exceptions caught during cleanup processing
+     * @return The same {@link LifeCycleStopFailureHandler} instance passed to this method
      */
-    private List<Throwable> stopInternal(boolean logFailuresImmediately)
+    public <T extends LifeCycleStopFailureHandler> T stop(T handler)
```

Comment:
After changing to lambda, we don't need this to return the handler -- it can be `void`. The caller already has the handler if they need it.